integration/container/attach_test.go

```diff
@@ -0,0 +1,206 @@
+// +build !windows
+
+package container
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/integration/internal/container"
+	"gotest.tools/assert"
+	is "gotest.tools/assert/cmp"
+	"gotest.tools/poll"
+	"gotest.tools/skip"
+)
+
+// #9860 Make sure attach ends when container ends (with no errors)
+func TestAttachClosedOnContainerStop(t *testing.T) {
+	skip.If(t, testEnv.IsRemoteDaemon, "test requires daemon on the same host")
+
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	cID := container.Run(t, ctx, client, container.WithTty(true),
+		container.WithCmd("/bin/sh", "-c", `trap 'exit 0' SIGTERM; while true; do sleep 1; done`),
+		func(c *container.TestContainerConfig) {
+			c.Config.OpenStdin = true
+		},
+	)
+	defer client.ContainerRemove(context.Background(), cID, types.ContainerRemoveOptions{Force: true})
+	poll.WaitOn(t, container.IsInState(ctx, client, cID, "running"), poll.WithDelay(5*time.Second))
+
+	attach, err := client.ContainerAttach(ctx, cID,
+		types.ContainerAttachOptions{
+			Stream: true,
+			Stdin:  true,
+			Stdout: true,
+			Stderr: true,
+		},
+	)
+	assert.NilError(t, err)
+	defer attach.Close()
+
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+		time.Sleep(300 * time.Millisecond)
+		err = client.ContainerStop(ctx, cID, nil)
+		assert.NilError(t, err)
+		_, err = io.Copy(&bytes.Buffer{}, attach.Reader)
+		errCh <- err
+	}()
+
+	poll.WaitOn(t, container.IsInState(ctx, client, cID, "exited"), poll.WithDelay(5*time.Second))
+
+	select {
+	case err = <-errCh:
+		assert.NilError(t, err)
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out waiting for attach channel to be closed")
+	}
+}
+
+func TestAttachAfterDetach(t *testing.T) {
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	cID := container.Run(t, ctx, client, container.WithTty(true),
+		container.WithCmd("sh"),
+		func(c *container.TestContainerConfig) {
+			c.Config.OpenStdin = true
+		},
+	)
+
+	defer client.ContainerRemove(context.Background(), cID, types.ContainerRemoveOptions{Force: true})
```

Comment:
Same here; I think the `defer setupTest(t)()` should already take care of this (if it isn't, then we need to fix that)

Reply:
Thanks, Updated the PR.